### PR TITLE
binarytrees: fix depth condition and optimize

### PR DESCRIPTION
**Pull request by cristicbz**
This fixes the `depth >= 0` vs `depth > 0` allocation issue, but that pushed the speed up to 2s on my machine (where the C version does around 1.4s-1.5s).

I optimised a bunch of stuff which brings the minimum runtimes down to about the same as C, but with higher variance (sometimes it'll take 1.7s).

I'm guessing this is because we allocate one arena per iteration, rather than reusing the arena. I think `typed_arena` should provide a `fn recycle(self) -> Self` method, which reuses allocations (taking `self` by value to ensure no outstanding references exist).

### CODE REVIEW

**Comment by TeXitoi** [comment]
Oups, forgot about this PR :( I've done conflicting modifications, sorry.

**Comment by cristicbz** [comment]
Ok, I resolved the conflicts, this still runs about 300ms faster on my machine (from 1.7s down to 1.4s).

File: src/binary_trees.rs [hunk 1]
```diff
@@ -13,26 +13,42 @@ use typed_arena::Arena;
 use rayon::prelude::*;
 
 struct Tree<'a> {
-    l: Option<&'a Tree<'a>>,
-    r: Option<&'a Tree<'a>>,
-    i: i32,
+    children: Option<(&'a Tree<'a>, &'a Tree<'a>)>,
```
Comment (cristicbz) [inline src/binary_trees.rs:16]: A node has either two children or none, this saves half the null-checks in the matching code below.

File: src/binary_trees.rs [hunk 2]
```diff
@@ -13,26 +13,42 @@ use typed_arena::Arena;
...
-fn bottom_up_tree<'r>(arena: &'r Arena<Tree<'r>>, i: i32, depth: i32) -> Option<&'r Tree<'r>> {
-    let (l, r) = if depth > 0 {
-        (bottom_up_tree(arena, 2 * i - 1, depth - 1), bottom_up_tree(arena, 2 * i, depth - 1))
-    } else {
-        (None, None)
-    };
-    Some(arena.alloc(Tree { l: l, r: r, i: i }))
+fn bottom_up_tree<'r>(arena: &'r Arena<Tree<'r>>, item: i32, depth: i32)
+                  -> &'r Tree<'r> {
+    let mut tree = arena.alloc(Tree { children: None, item: item });
```
Comment (cristicbz) [inline src/binary_trees.rs:46]: By allocating the node first, the arena is populated closer to iteration order: if we allocate the node at the end then the arena is first populated with all the leaves, and then with parents.

**Review by TeXitoi — APPROVED** [review]
Great!

**Comment by TeXitoi** [comment]
Thus, we're competitive with the C version now?

File: src/binary_trees.rs [hunk 3]
```diff
@@ -62,7 +78,7 @@ fn main() {
     let long_lived_arena = Arena::new();
     let long_lived_tree = bottom_up_tree(&long_lived_arena, 0, max_depth);
 
-    let messages = (min_depth/2..max_depth/2 + 1).into_par_iter().weight_max().map(|half_depth| {
+    let messages = (min_depth/2..max_depth/2 + 1).into_par_iter().map(|half_depth| {
```
Comment (cristicbz) [inline src/binary_trees.rs:81]: With these changes, empirically `weight_max` didn't help anymore.

**Comment by cristicbz** [comment]
@TeXitoi  I think so! But the mandelbrot failure with rayon makes me worry that for some reason Isaac's computer and rayon don't play well :/. Let's see!

**Comment by TeXitoi** [comment]
Did you bench without LTO? Looks like Isaac is not using LTO, it was causing some troubles some times ago.

**Comment by TeXitoi** [comment]
updated https://alioth.debian.org/tracker/index.php?func=detail&aid=315632&group_id=100815&atid=413122

**Comment by cristicbz** [comment]
No I hadn't tried without LTO, but I tried now, with no difference in perf.

**Pull request merged**